stdlib/digest.ml

```diff
@@ -13,10 +13,139 @@
 (*                                                                        *)
 (**************************************************************************)
 
-(* Message digest (MD5) *)
+(* Utility functions *)
+
+let hex_of_string d =
+  let char_hex n =
+    Char.chr (if n < 10 then Char.code '0' + n
+                        else Char.code 'a' + n - 10) in
+  let len = String.length d in
+  let result = Bytes.create (len * 2) in
+  for i = 0 to len - 1 do
+    let x = Char.code d.[i] in
+    Bytes.unsafe_set result (i*2) (char_hex (x lsr 4));
+    Bytes.unsafe_set result (i*2+1) (char_hex (x land 0x0f));
+  done;
+  Bytes.unsafe_to_string result
+
+let string_of_hex s =
+  let digit c =
+    match c with
+    | '0'..'9' -> Char.code c - Char.code '0'
+    | 'A'..'F' -> Char.code c - Char.code 'A' + 10
+    | 'a'..'f' -> Char.code c - Char.code 'a' + 10
+    | _ -> invalid_arg "Digest.from_hex" in
+  let byte i = digit s.[i] lsl 4 + digit s.[i+1] in
+  String.init (String.length s / 2) (fun i -> Char.chr (byte (2 * i)))
+
+(* Generic interface *)
+
+module type S = sig
+  type t = string
+  val hash_size : int
+  val compare : t -> t -> int
+  val equal : t -> t -> bool
+  val string : string -> t
+  val bytes : bytes -> t
+  val substring : string -> int -> int -> t
+  val subbytes : bytes -> int -> int -> t
+  val channel : in_channel -> int -> t
+  val file : string -> t
+  val output : out_channel -> t -> unit
+  val input : in_channel -> t
+  val to_hex : t -> string
+  val from_hex : string -> t
+end
+
+(* BLAKE2 hashing, parameterized by hash size *)
+
+module BLAKE2 (X: sig val hash_size : int end) : S = struct
+
+type t = string
+
+let hash_size =
+  if X.hash_size < 1 || X.hash_size > 64
+  then invalid_arg "Digest.BLAKE2: wrong hash size";
+  X.hash_size
+
+let compare = String.compare
+let equal = String.equal
+
+type state
+
+external create_gen: int -> string -> state = "caml_blake2_create"
+external update: state -> string -> int -> int -> unit = "caml_blake2_update"
+external final: state -> int -> t = "caml_blake2_final"
+external unsafe_string: int -> string -> string -> int -> int -> t
+                      = "caml_blake2_string"
```

Comment:
What about tweaking that to a `HASH` interface without the blake specific `int`s. And provide `Digest.Make` that derives `S` from that. This could be useful for the larger eco-system.

Reply:
Currently, the implementation is structured like this, with a `Make` functor parameterized by a low-level hash function, but I haven't exported it yet, and I'm not sure it would really be usable.